File: leetcode/facebook/binary-tree-maximum-path-sum.py
"""
https://leetcode.com/problems/binary-tree-maximum-path-sum
"""


# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.res = 0 - 2 ** 31


        def _helper(node):
            if not node:
                return 0

            sum_l, sum_r = max(0, _helper(node.left)), max(0, _helper(node.right))
            sum_n = node.val + sum_l + sum_r
            if sum_n > self.res:
                self.res = sum_n
            return node.val + max(sum_l, sum_r)

        _helper(root)
        return self.res


    def maxPathSum_2(self, root):
        self.res = 0 - 2 ** 31
        self.path = None

        def _helper(node):
            if not node:
                return 0, None
            sum_l, path_l = _helper(node.left)
            if sum_l <= 0:
                sum_l = 0
                path_l = None
            sum_r, path_r = _helper(node.right)
            if sum_r <= 0:
                sum_r = 0
                path_r = None
            sum_n = node.val + sum_l + sum_r
            if sum_n > self.res:
                self.res = sum_n
                self.path = TreeNode(node.val)
                self.path.left, self.path.right = path_l, path_r
            path_n = TreeNode(node.val)
            if sum_l > sum_r:
                path_n.left = path_l
                return node.val + sum_l, path_n
            else:
                path_n.right = path_r
                return node.val + sum_r, path_n

        _helper(root)
        return self.path

root = TreeNode(-5)
n2 = TreeNode(-2)
n3 = TreeNode(-3)
n4 = TreeNode(-2)
root.left, root.right = n2, n3
n2.left = n4
print(Solution().maxPathSum_2(root))

File: leetcode/go/binary_tree_maximum_path_sum.go
package main

import (
	"math"
	"fmt"
)

/*
https://leetcode.com/problems/binary-tree-maximum-path-sum

For each node N in the tree, we could calculate the sum of the tree with N as
the root. And check the sum with the current max one to see whether it could be
the new max one. Then recursively, N's longer path (right or left) could be in
path of N's parent with max sum.
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func max(a int, b int) int {
	if a > b {
		return a
	}
	return b
}

func maxPathSum(root *TreeNode) int {
	res := math.MinInt32
	var helper func(*TreeNode) int
	helper = func(node *TreeNode) int {
		if node == nil {
			return 0
		}
		sumLeft, sumRight := max(0, helper(node.Left)), max(0, helper(node.Right))
		sum := node.Val + sumLeft + sumRight
		res = max(res, sum)
		return node.Val + max(sumLeft, sumRight)
	}

	helper(root)
	return res
}

func main() {
	n1 := &TreeNode{1, nil, nil}
	n2 := &TreeNode{-2, nil, nil}
	n3 := &TreeNode{3, nil, nil}
	n4 := &TreeNode{4, nil, nil}
	n5 := &TreeNode{5, nil, nil}

	n3.Left = n2
	n3.Right = n5
	n5.Left = n1
	n5.Right = n4

	fmt.Println(maxPathSum(n3))
}
